src/ion/liveranges.rs

```diff
@@ -0,0 +1,1192 @@
+/*
+ * The following license applies to this file, which was initially
+ * derived from the files `js/src/jit/BacktrackingAllocator.h` and
+ * `js/src/jit/BacktrackingAllocator.cpp` in Mozilla Firefox:
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * Since the initial port, the design has been substantially evolved
+ * and optimized.
+ */
+
+//! Live-range computation.
+
+use super::{
+    CodeRange, Env, InsertMovePrio, LiveBundle, LiveBundleIndex, LiveRange, LiveRangeFlag,
+    LiveRangeIndex, LiveRangeKey, LiveRangeListEntry, LiveRangeSet, PRegData, PRegIndex, RegClass,
+    SpillSetIndex, Use, VRegData, VRegIndex, SLOT_NONE,
+};
+use crate::bitvec::BitVec;
+use crate::{
+    Allocation, Block, Function, Inst, InstPosition, Operand, OperandKind, OperandPolicy,
+    OperandPos, PReg, ProgPoint, RegAllocError, VReg,
+};
+use fxhash::FxHashSet;
+use smallvec::{smallvec, SmallVec};
+use std::collections::{HashSet, VecDeque};
+use std::convert::TryFrom;
+
+#[inline(always)]
+pub fn spill_weight_from_policy(policy: OperandPolicy, loop_depth: usize, is_def: bool) -> u32 {
+    // A bonus of 1000 for one loop level, 4000 for two loop levels,
+    // 16000 for three loop levels, etc. Avoids exponentiation.
+    let hot_bonus = std::cmp::min(16000, 1000 * (1 << (2 * loop_depth)));
```

Comment:
I'm getting overflows on the multiplication here with `loop_depth = 12`. Also the shift can potentially overflow when `loop_depth >= 16`.

Note that I haven't checked my IR to see if the loop depth is actually 12 or if the depth count is getting confused by irreducible control flow.

Reply:
Thanks! Just pushed a commit that should fix this.